[ci] Update macos runner version

SukkaW opened the pull request:
<!--
Thank you for your contribution!

If it is your first time contributing to Taichi, please read our Contributor Guidelines:
  https://docs.taichi-lang.org/docs/contributor_guide

- Please always prepend your PR title with tags such as [CUDA], [Lang], [Doc], [Example]. For a complete list of valid PR tags, please check out https://github.com/taichi-dev/taichi/blob/master/misc/prtags.json.
- Use upper-case tags (e.g., [Metal]) for PRs that change public APIs. Otherwise, please use lower-case tags (e.g., [metal]).
- More details: https://docs.taichi-lang.org/docs/contributor_guide#pr-title-format-and-tags

- Please fill in the issue number that this PR relates to.
- If your PR fixes the issue **completely**, use the `close` or `fixes` prefix so that GitHub automatically closes the issue when the PR is merged. For example,
    Related issue = close #2345
- If the PR does not belong to any existing issue, free to leave it blank.
-->

GitHub Action is sunsetting the macOS 10.15 Actions runner. It will stop working intermittently until being completely removed by 2022-8-30: https://github.blog/changelog/2022-07-20-github-actions-the-macos-10-15-actions-runner-image-is-being-deprecated-and-will-be-removed-by-8-30-22/

Note: the `CI_PLATFORM` env variable is used to download the prebuilt version of LLVM, and setting it to `macos` is completely ok.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/taichi-dev/taichi?pullRequest=5505) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/taichi-dev/taichi?pullRequest=5505) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/taichi-dev/taichi?pullRequest=5505) it.</sub>

feisuzhu commented:
Hi, I think it didn't work out :(

`actions/setup-python` is for github runners so we have to limit the task not to run on self-hosted runners.

feisuzhu commented:
Hi, since taichi needs to support macOS 10.15, I moved the CPU build task to internal buildbot, but we also need to build & test on latest macOS, so could you add a new task addressing this (if you are still interested)?

SukkaW commented:
> Hi, since taichi needs to support macOS 10.15, I moved the CPU build task to internal buildbot, but we also need to build & test on latest macOS, so could you add a new CI task addressing this (if you are still interested)?

Let me see what I can do on this.

SukkaW commented:
I have rebased the PR to the latest `master`, and modify the `strategy` to support multiple `runs-on`.

SukkaW commented on .github/workflows/release.yml:
```diff
@@ -200,10 +200,12 @@ jobs:
       fail-fast: false
-      matrix: ${{ fromJson(needs.matrix_prep.outputs.matrix) }}
-    runs-on: [self-hosted, macos-10.15]
+      matrix:
+       from_matrix_prep: ${{ fromJson(needs.matrix_prep.outputs.matrix) }}
+       runs_on: [[self-hosted, macos-10.15], macos-latest]
```
GitHub Action only supports multiple `runs-on` using `strategy`. In order to minimize the changes, I change the original `matrix` to `matrix.from_matrix_prep` and `matrix.runs_on`.

However, I don't know if GitHub Action even recognizes `[[self-hosted, macos-10.15], macos-latest]`.

feisuzhu replied:
Hi, sorry for the late reply, the notification is somehow lost for no reason 😅

macOS are generally backward compatible so we don't need to build a whole matrix of wheels here (with 10.14 and the M1 chip transition as exceptions, which already addressed by `build_and_test_1014` and `build_and_test_m1` task respectively), but we do need to ensure built wheels are behaving normally on newer systems, so IMHO these should be moved to `testing.yml`

feisuzhu commented on .github/workflows/testing.yml:
```diff
@@ -190,7 +190,7 @@ jobs:
       matrix:
         include:
-          - os: macos-10.15
+          - os: macos-latest
```
Both `10.15` and `latest` should be tested

SukkaW commented:
@feisuzhu I have changed the `testing` to run on `[[self-hosted, macos-10.15], macos-latest]` (I am still not sure if this syntax would work. If not, we might need a separate task), while the `release` is now only run on `[self-hosted, macos-10.15]`.

I have also rebased the PR to the latest `master` branch.

feisuzhu commented:
@SukkaW No, it's not working 😵

Pasting error message here in case you can't see it
![2022-08-15-161203_942x246_scrot](https://user-images.githubusercontent.com/857880/184600013-d13f11f9-f713-4226-9087-ac7ae2f5af16.png)

feisuzhu commented:
I'd recommend you to tinker it in your own repo to avoid this extra long trial-and-error process.

SukkaW commented:
> Pasting error message here in case you can't see it ![2022-08-15-161203_942x246_scrot]
> I'd recommend you to tinker it in your own repo to avoid this extra long trial-and-error process.

I do not have a self-hosted github action runner thus I can't tinker with it in my own repo (the process will fail anyway).

The error message indicates that GitHub Action is not able to parse `[[self-hosted, macos-10.15], macos-latest]`. So here are two solutions:

- Switch to the YAML array to see if GitHub Action can identify it
- Split it into 2 separate tasks with different `runs-on`

feisuzhu commented on .github/workflows/testing.yml:
```diff
@@ -208,14 +208,12 @@ jobs:
       matrix:
         include:
-          - os: macos-10.15
+          - os: [[self-hosted, macos-10.15], macos-latest]
             python: 3.7
             with_cc: OFF
             with_cpp_tests: ON
             wanted_archs: 'cpu'
```
```suggestion
          - os: [self-hosted, macos-10.15]
            python: 3.7
            with_cc: OFF
            with_cpp_tests: ON
            wanted_archs: 'cpu'
          - os: macos-latest
            python: 3.7
            with_cc: OFF
            with_cpp_tests: ON
            wanted_archs: 'cpu'
```

Could you please try this?
It might not work though since github hosted runners may not have brew.

feisuzhu commented:
Closing this PR since it no longer applies to current situation.
